File: test/Dapr.IntegrationTest.AI/Conversation/ConversationTests.cs
// ------------------------------------------------------------------------
// Copyright 2025 The Dapr Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------

using Dapr.AI.Conversation;
using Dapr.AI.Conversation.ConversationRoles;
using Dapr.AI.Conversation.Extensions;
using Dapr.Testcontainers;
using Dapr.Testcontainers.Common;
using Dapr.Testcontainers.Harnesses;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;

namespace Dapr.IntegrationTest.AI.Conversation;

public sealed class ConversationTests
{
    [Fact]
    public async Task ShouldProcessConversation()
    {
        var componentsDir = TestDirectoryManager.CreateTestDirectory("conversation-components");

        await using var environment = await DaprTestEnvironment.CreateWithPooledNetworkAsync();
        await environment.StartAsync();

        var harness = new DaprHarnessBuilder(componentsDir)
            .WithEnvironment(environment)
            .BuildConversation();
        await using var testApp = await DaprHarnessBuilder.ForHarness(harness)
            .ConfigureServices(builder =>
            {
                builder.Services.AddDaprConversationClient((sp, clientBuilder) =>
                {
                    var config = sp.GetRequiredService<IConfiguration>();
                    var grpcEndpoint = config["DAPR_GRPC_ENDPOINT"];
                    var httpEndpoint = config["DAPR_HTTP_ENDPOINT"];

                    if (!string.IsNullOrEmpty(grpcEndpoint))
                        clientBuilder.UseGrpcEndpoint(grpcEndpoint);
                    if (!string.IsNullOrEmpty(httpEndpoint))
                        clientBuilder.UseHttpEndpoint(httpEndpoint);
                });
            })
            .BuildAndStartAsync();

        using var scope = testApp.CreateScope();
        var client = scope.ServiceProvider.GetRequiredService<DaprConversationClient>();

        var inputs = new[]
        {
            new ConversationInput(
            [
                new SystemMessage
                    {
                        Content = [new MessageContent("You are a concise assistant.")]
                    },
                    new UserMessage
                    {
                        Content = [new MessageContent("Respond with a short greeting.")]
                    }
            ])
        };

        var options = new ConversationOptions(Constants.DaprComponentNames.ConversationComponentName)
        {
            Temperature = 0
        };

        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(60));
        var response = await client.ConverseAsync(inputs, options, cts.Token);

        Assert.NotNull(response);
        Assert.Single(response.Outputs);
        Assert.NotEmpty(response.Outputs[0].Choices);
        Assert.False(string.IsNullOrWhiteSpace(response.Outputs[0].Choices[0].Message.Content));
    }
}

File: test/Dapr.IntegrationTest.FSharp.AI/ConversationTests.fs
#nowarn "FS3261"
#nowarn "57"

namespace Dapr.IntegrationTest.FSharp.AI

open System
open System.Threading
open Dapr.AI.Conversation
open Dapr.AI.Conversation.ConversationRoles
open Dapr.AI.Conversation.Extensions
open Dapr.AI.Conversation.Tools
open Dapr.Testcontainers
open Dapr.Testcontainers.Common
open Dapr.Testcontainers.Harnesses
open global.Dapr.Testcontainers.Xunit.Attributes
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open global.Xunit

type ConversationTests() =
    [<MinimumDaprRuntimeFact("1.17")>]
    member _.ShouldProcessConversation() = task {
        let componentsDir = TestDirectoryManager.CreateTestDirectory("fsharp-conversation-components")

        use! env = DaprTestEnvironment.CreateWithPooledNetworkAsync(cancellationToken = TestContext.Current.CancellationToken)
        do! env.StartAsync(TestContext.Current.CancellationToken)

        let harness = DaprHarnessBuilder(componentsDir).WithEnvironment(env).BuildConversation()
        use! testApp =
            DaprHarnessBuilder.ForHarness(harness)
                .ConfigureServices(fun builder ->
                    builder.Services.AddDaprConversationClient(
                        Action<IServiceProvider, DaprConversationClientBuilder>(
                            fun sp clientBuilder ->
                                let config = sp.GetRequiredService<IConfiguration>()
                                let grpcEndpoint = config["DAPR_GRPC_ENDPOINT"]
                                let httpEndpoint = config["DAPR_HTTP_ENDPOINT"]
                                if not (String.IsNullOrEmpty(grpcEndpoint)) then
                                    clientBuilder.UseGrpcEndpoint(grpcEndpoint) |> ignore
                                if not (String.IsNullOrEmpty(httpEndpoint)) then
                                    clientBuilder.UseHttpEndpoint(httpEndpoint) |> ignore))
                    |> ignore)
                .BuildAndStartAsync()

        use scope = testApp.CreateScope()
        let client = scope.ServiceProvider.GetRequiredService<DaprConversationClient>()

        let inputs =
            [|
                ConversationInput(
                    [|
                        SystemMessage(
                            Content = [| MessageContent("You are a concise assistant.") |]
                        )
                        UserMessage(
                            Content = [| MessageContent("Respond with a short greeting.") |]
                        )
                    |])
            |]

        let options =
            ConversationOptions(Constants.DaprComponentNames.ConversationComponentName,
                Temperature = Nullable(0.0))

        use cts = new CancellationTokenSource(TimeSpan.FromSeconds(60.0))
        let! response = client.ConverseAsync(inputs, options, cts.Token)

        Assert.NotNull(response)
        Assert.Single(response.Outputs) |> ignore
        Assert.NotEmpty(response.Outputs[0].Choices)
        Assert.False(String.IsNullOrWhiteSpace(response.Outputs[0].Choices[0].Message.Content))
    }
